kubernetes/csi/lib/ubi_csi/node_service.rb

```diff
@@ -0,0 +1,359 @@
+require "base64"
+require_relative "errors"
+require_relative "kubernetes_client"
+require_relative "../csi_services_pb"
+
+module Csi
+  module V1
+    class NodeService < Node::Service
+      include Csi::ServiceHelper
+
+      MAX_VOLUMES_PER_NODE = 8
+      VOLUME_BASE_PATH = "/var/lib/ubicsi"
+      OLD_PV_NAME_ANNOTATION_KEY = "csi.ubicloud.com/old-pv-name"
+      OLD_PVC_OBJECT_ANNOTATION_KEY = "csi.ubicloud.com/old-pvc-object"
+
+      def initialize(logger:, node_id:)
+        FileUtils.mkdir_p(VOLUME_BASE_PATH) unless Dir.exist?(VOLUME_BASE_PATH)
+        @logger = logger
+        @node_id = node_id
+      end
+
+      attr_reader :node_id
+
+      def node_get_capabilities(req, _call)
+        log_request_response(req, "node_get_capabilities") do |req_id|
+          NodeGetCapabilitiesResponse.new(
+            capabilities: [
+              NodeServiceCapability.new(
+                rpc: NodeServiceCapability::RPC.new(
+                  type: NodeServiceCapability::RPC::Type::STAGE_UNSTAGE_VOLUME
+                )
+              )
+            ]
+          )
+        end
+      end
+
+      def node_get_info(req, _call)
+        log_request_response(req, "node_get_info") do |req_id|
+          topology = Topology.new(
+            segments: {
+              "kubernetes.io/hostname" => @node_id
+            }
+          )
+          NodeGetInfoResponse.new(
+            node_id: @node_id,
+            max_volumes_per_node: MAX_VOLUMES_PER_NODE,
+            accessible_topology: topology
+          )
+        end
+      end
+
+      def run_cmd_output(*cmd, req_id:)
+        output, _ = run_cmd(*cmd, req_id:)
+        output
+      end
+
+      def is_mounted?(path, req_id:)
+        _, status = run_cmd("mountpoint", "-q", path, req_id:)
+        status == 0
+      end
+
+      def find_loop_device(backing_file, req_id:)
+        output, ok = run_cmd("losetup", "-j", backing_file, req_id:)
+        if ok && !output.empty?
+          loop_device = output.split(":", 2)[0].strip
+          return loop_device if loop_device.start_with?("/dev/loop")
+        end
+        nil
+      end
+
+      def self.backing_file_path(volume_id)
+        File.join(VOLUME_BASE_PATH, "#{volume_id}.img")
+      end
+
+      def pvc_needs_migration?(pvc)
+        old_pv_name = pvc.dig("metadata", "annotations", OLD_PV_NAME_ANNOTATION_KEY)
+        !old_pv_name.nil?
+      end
+      alias_method :is_copied_pvc?, :pvc_needs_migration?
+
+      def node_stage_volume(req, _call)
+        log_request_response(req, "node_stage_volume") do |req_id|
+          client = KubernetesClient.new(req_id:, logger: @logger)
+          pvc = fetch_and_migrate_pvc(req_id, client, req)
+          perform_node_stage_volume(req_id, pvc, req, _call)
+          roll_back_reclaim_policy(req_id, client, req, pvc)
+          remove_old_pv_annotation(client, pvc)
+          NodeStageVolumeResponse.new
+        end
+      end
+
+      def fetch_and_migrate_pvc(req_id, client, req)
```

Comment:
I believe in-code documentation would be super useful, especially for this file. But it's not a blocker for the PR and can be followed up later.